<a href="https://colab.research.google.com/github/chrisjallaine/CloudComputingDataEngineeringActivity/blob/main/ETL_Phase_Data_Warehouse.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Dependencies

In [1]:
import pandas as pd
import sqlalchemy
import numpy as np
from scipy import stats

## Load the Data

In [2]:
# connect to the database
conn = sqlalchemy.create_engine('postgresql://dbaas_menorias_user:Rg7E7gY3B0s42eqKOCaWxMOqGiL0ZWKY@dpg-d05hb0juibrs73fmp600-a.oregon-postgres.render.com/dbaas_menorias')

# read the data from the table into a DataFrame
data = pd.read_sql('SELECT * FROM hardware_sales', conn)

## Data Information

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89833 entries, 0 to 89832
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   sls_ord_num         89833 non-null  object 
 1   prd_key             89833 non-null  object 
 2   CID                 89833 non-null  int64  
 3   sls_order_dt        89833 non-null  int64  
 4   sls_ship_dt         89833 non-null  int64  
 5   sls_due_dt          89833 non-null  int64  
 6   sls_sales           89818 non-null  float64
 7   sls_quantity        89833 non-null  int64  
 8   sls_price           89822 non-null  float64
 9   BDATE               89833 non-null  object 
 10  GEN                 83450 non-null  object 
 11  cst_id              89833 non-null  float64
 12  cst_firstname       89823 non-null  object 
 13  cst_lastname        89822 non-null  object 
 14  cst_marital_status  89826 non-null  object 
 15  cst_gndr            58500 non-null  object 
 16  cst_

## Metadata

# Dataset Metadata

| Column Name            | Non-Null Count | Data Type | Description (Inferred)                     | Notes                              |
|------------------------|----------------|-----------|--------------------------------------------|------------------------------------|
| **sls_ord_num**        | 89,833         | `object`  | Sales order number (unique identifier)     | Likely alphanumeric.               |
| **prd_key**           | 89,833         | `object`  | Product key (identifier)                   | Foreign key for products.          |
| **CID**               | 89,833         | `int64`   | Customer ID                                | Unique customer identifier.        |
| **sls_order_dt**      | 89,833         | `int64`   | Order date (numeric format)                | Needs conversion to datetime.      |
| **sls_ship_dt**       | 89,833         | `int64`   | Shipment date (numeric format)             | Needs conversion to datetime.      |
| **sls_due_dt**        | 89,833         | `int64`   | Due date (numeric format)                  | Needs conversion to datetime.      |
| **sls_sales**         | 89,818         | `float64` | Total sales amount                         | 15 missing values.                 |
| **sls_quantity**      | 89,833         | `int64`   | Quantity of products sold                  |                                    |
| **sls_price**         | 89,822         | `float64` | Unit price of product                      | 11 missing values.                 |
| **BDATE**             | 89,833         | `object`  | Birthdate (string format)                  | Needs datetime parsing.            |
| **GEN**               | 83,450         | `object`  | Generation (e.g., Gen X, Millennial)       | 6,383 missing values.              |
| **cst_id**            | 89,833         | `float64` | Customer ID (alternative)                  | Duplicate of `CID`?                |
| **cst_firstname**     | 89,823         | `object`  | Customer first name                        | 10 missing values.                 |
| **cst_lastname**      | 89,822         | `object`  | Customer last name                         | 11 missing values.                 |
| **cst_marital_status**| 89,826         | `object`  | Marital status (e.g., Single, Married)     | 7 missing values.                  |
| **cst_gndr**          | 58,500         | `object`  | Customer gender                            | 31,333 missing values (35%).       |
| **cst_create_date**   | 89,833         | `object`  | Customer account creation date             | Needs datetime parsing.            |
| **CNTRY**             | 88,484         | `object`  | Customer country                           | 1,349 missing values.              |
| **prd_id**            | 89,833         | `int64`   | Product ID                                 | Unique identifier.                 |
| **prd_nm**            | 89,833         | `object`  | Product name                               |                                    |
| **prd_cost**          | 89,833         | `float64` | Product cost                               |                                    |
| **prd_line**          | 89,833         | `object`  | Product line/category                      | e.g., Electronics, Apparel.        |
| **prd_start_dt**      | 89,833         | `object`  | Product launch date                        | Needs datetime parsing.            |
| **prd_end_dt**        | 32,667         | `object`  | Product discontinuation date               | 57,166 missing (likely active).    |
| **ID**                | 89,833         | `object`  | Additional identifier (ambiguous)          | Check business context.            |
| **CAT**               | 89,833         | `object`  | Product category                           | e.g., Hardware, Software.          |
| **SUBCAT**            | 89,833         | `object`  | Product subcategory                        |                                    |
| **MAINTENANCE**       | 89,833         | `object`  | Maintenance flag (e.g., Yes/No)            | Binary or descriptive?             |

### Key Observations:
1. **Temporal Columns**:  
   - Dates (`sls_order_dt`, `sls_ship_dt`, `BDATE`, etc.) are stored as `int64` or `object`—need conversion to datetime.
2. **Missing Values**:  
   - `cst_gndr` (35% missing), `prd_end_dt` (64% missing), `GEN` (7% missing).
3. **Potential Duplicates**:  
   - `CID` (int) vs. `cst_id` (float)—verify if they represent the same entity.
4. **Product Lifecycle**:  
   - `prd_end_dt` missing for most entries, suggesting active products.
5. **Sales Metrics**:  
   - `sls_sales`, `sls_price` have minimal missing values (<0.02%).

## Data Cleaning

### Convert Date Columns

- Columns stored as int64 or object need conversion to datetime

In [4]:
date_columns = {
    'sls_order_dt': 'order_date',  # Rename for clarity
    'sls_ship_dt': 'ship_date',
    'sls_due_dt': 'due_date',
    'BDATE': 'birth_date',
    'cst_create_date': 'customer_creation_date',
    'prd_start_dt': 'product_start_date',
    'prd_end_dt': 'product_end_date'
}

for col, new_name in date_columns.items():
    data[new_name] = pd.to_datetime(data[col], errors='coerce')
    data.drop(col, axis=1, inplace=True)

## Handle Missing Values

### High-Missing Columns:
*cst_gndr* (35% missing):

- Impute with "Unknown" or use predictive modeling if gender is critical.

*prd_end_dt* (64% missing):

- Missing likely indicates active products. Create a flag.

In [5]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

In [6]:
data['is_active_product'] = data['product_end_date'].isna().astype(int)

## Low-Missing Columns

*sls_sales, sls_price, cst_firstname*, etc.:

- Drop rows or impute with median/mean (numeric) or mode (categorical):



In [7]:
numeric_cols = ['sls_sales', 'sls_price', 'prd_cost']
data[numeric_cols] = data[numeric_cols].fillna(data[numeric_cols].median())

categorical_cols = ['cst_firstname', 'cst_lastname', 'cst_marital_status']
data[categorical_cols] = data[categorical_cols].fillna('Not Specified')

## Standardize Categorical Data

*GEN* (Generation), *CNTRY* (Country), *prd_line*:

- Convert to uppercase/lowercase and trim whitespace

In [8]:
data['GEN'] = data['GEN'].str.upper().str.strip()
data['CNTRY'] = data['CNTRY'].str.upper().str.strip()

*cst_gndr*

- Standardize values (e.g., "M", "F", "Unknown")

In [9]:
gender_map = {'Male': 'M', 'Female': 'F', 'm': 'M', 'f': 'F'}
data['cst_gndr'] = data['cst_gndr'].replace(gender_map).fillna('Unknown')

## Deduplicate and Validate Identifiers

Check *CID* vs. *cst_id*

In [10]:
if data['CID'].equals(data['cst_id'].astype('int64')):
    data.drop('cst_id', axis=1, inplace=True)  # Remove duplicate

## Validate sls_ord_num (Sales Order Number)

- Ensure uniqueness

In [11]:
duplicate_count = len(data) - data['sls_ord_num'].nunique()
print(f"{duplicate_count} duplicate order numbers found.")

62174 duplicate order numbers found.


In [12]:
data = data.drop_duplicates(subset='sls_ord_num', keep='first')

## Handle Product Data
*Active* vs. *Discontinued* Products:

- Use product_end_date to flag status.

### Product Categories (CAT, SUBCAT):

- Merge or split if hierarchies are unclear:

In [13]:
data['full_category'] = data['CAT'] + ' > ' + data['SUBCAT']

## Fix Numeric Outliers
*sls_quantity*, *sls_price*:

- Clip extreme values (e.g., beyond 3 standard deviations)

In [14]:
data = data[(np.abs(stats.zscore(data[['sls_quantity', 'sls_price']])) < 3).all(axis=1)]

## Feature Engineering

In [15]:
# Customer Age
data['customer_age'] = (pd.to_datetime('today') - data['birth_date']).dt.days // 365

In [16]:
# Order-to-Ship Delay
data['order_to_ship_days'] = (data['ship_date'] - data['order_date']).dt.days

## Validation

In [17]:
data.isna().sum()

sls_ord_num                   0
prd_key                       0
CID                           0
sls_sales                     0
sls_quantity                  0
sls_price                     0
GEN                        1820
cst_firstname                 0
cst_lastname                  0
cst_marital_status            0
cst_gndr                      0
CNTRY                       332
prd_id                        0
prd_nm                        0
prd_cost                      0
prd_line                      0
ID                            0
CAT                           0
SUBCAT                        0
MAINTENANCE                   0
order_date                    0
ship_date                     0
due_date                      0
birth_date                    9
customer_creation_date        0
product_start_date            0
product_end_date          17088
is_active_product             0
full_category                 0
customer_age                  9
order_to_ship_days            0
dtype: int64

In [18]:
data = data.dropna()

In [19]:
data.head()

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
1     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
2     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
3     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
5     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
6     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   

  cst_firstname cst_lastname cst_marital_status  ...  \
1       Rachael     Martinez                  S  ...   
2        Sydney       Wright                  S  ...   
3         Ruben       Prasad                  M  ...   
5       Christy          Zhu                  S  ...   
6         Colin        Anand                  M  ...   

                      ship_date                      due_date  birth_date  \
1 1970-01-01 00:00:00.020110105 1970-01-01 00:00:00.020110110  1970-06-17   
2 1970-01-01 00:00:00.020110105 1970-01-01 00:00:00.020110110  1952-06-01   
3 1970-01-01 00:00:00.020110105 1970-01-01 00:00:00.020110110  1943-11-10   
5 1970-01-01 00:00:00.020110105 1970-01-01 00:00:00.020110110  1973-08-14   
6 1970-01-01 00:00:00.020110106 1970-01-01 00:00:00.020110111  1972-08-14   

  customer_creation_date  product_start_date product_end_date  \
1             2026-01-25          2011-07-01       2007-12-28   
2             2026-01-14          2011-07-01       2007-12-28   
3             2025-10-12          2011-07-01       2007-12-28   
5             2025-10-06          2011-07-01       2007-12-28   
6             2026-01-25          2011-07-01       2007-12-28   

  is_active_product           full_category customer_age order_to_ship_days  
1                 0  Bikes > Mountain Bikes         54.0                  0  
2                 0  Bikes > Mountain Bikes         73.0                  0  
3                 0      Bikes > Road Bikes         81.0                  0  
5                 0  Bikes > Mountain Bikes         51.0                  0  
6                 0      Bikes > Road Bikes         52.0                  0  

[5 rows x 31 columns]

## Sending to Data Warehouse

In [30]:
!pip install pandas sqlalchemy psycopg2-binary

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 29.4 MB/s eta 0:00:00


In [20]:
#Render database details
DATABASE_URL = "postgresql://user:67iAcybMHCCi7x7ASxxABEY55wsWM9KC@dpg-d0ou84be5dus73d9fgng-a.singapore-postgres.render.com/salesdashboard"

In [21]:
from sqlalchemy import create_engine

# Create a SQLAlchemy engine
engine = create_engine(DATABASE_URL)

# Export DataFrame to PostgreSQL
data.to_sql(
    name='sales_data',
    con=engine,
    if_exists='replace',
    index=False,
    chunksize=1000
)

print("Data successfully exported to Render PostgreSQL!")

Data successfully exported to Render PostgreSQL!


In [23]:
# Read back data to confirm
df_loaded = pd.read_sql('SELECT * FROM sales_data LIMIT 10', con=engine)
df_loaded

sls_ord_num     prd_key    CID  sls_sales  sls_quantity  sls_price     GEN  \
0     SO43698  BK-M82S-44  28389     3400.0             1     3400.0  FEMALE   
1     SO43699  BK-M82S-44  25863     3400.0             1     3400.0  FEMALE   
2     SO43700  BK-R50B-62  14501      699.0             1      699.0    MALE   
3     SO43701  BK-M82S-44  11003     3400.0             1     3400.0  FEMALE   
4     SO43702  BK-R93R-44  27645     3578.0             1     3578.0    MALE   
5     SO43703  BK-R93R-62  16624     3578.0             1     3578.0    MALE   
6     SO43704  BK-M82B-48  11005     3375.0             1     3375.0    MALE   
7     SO43705  BK-M82S-38  11011     3400.0             1     3400.0    MALE   
8     SO43706  BK-R93R-48  27621     3578.0             1     3578.0    MALE   
9     SO43707  BK-R93R-48  27616     3578.0             1     3578.0  FEMALE   

  cst_firstname cst_lastname cst_marital_status  ...  \
0       Rachael     Martinez                  S  ...   
1        Sydney       Wright                  S  ...   
2         Ruben       Prasad                  M  ...   
3       Christy          Zhu                  S  ...   
4         Colin        Anand                  M  ...   
5        Albert      Alvarez                  S  ...   
6         Julio         Ruiz                  S  ...   
7        Curtis           Lu                  M  ...   
8        Edward        Brown                  S  ...   
9          Emma        Brown                  S  ...   

                   ship_date                   due_date  birth_date  \
0 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1970-06-17   
1 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1952-06-01   
2 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1943-11-10   
3 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1973-08-14   
4 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1972-08-14   
5 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1983-07-24   
6 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1976-08-01   
7 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1969-05-03   
8 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1975-01-23   
9 1970-01-01 00:00:00.020110 1970-01-01 00:00:00.020110  1971-06-10   

  customer_creation_date  product_start_date product_end_date  \
0             2026-01-25          2011-07-01       2007-12-28   
1             2026-01-14          2011-07-01       2007-12-28   
2             2025-10-12          2011-07-01       2007-12-28   
3             2025-10-06          2011-07-01       2007-12-28   
4             2026-01-25          2011-07-01       2007-12-28   
5             2026-01-04          2011-07-01       2007-12-28   
6             2025-10-06          2011-07-01       2007-12-28   
7             2025-10-06          2011-07-01       2007-12-28   
8             2026-01-25          2011-07-01       2007-12-28   
9             2026-01-21          2011-07-01       2007-12-28   

  is_active_product           full_category customer_age order_to_ship_days  
0                 0  Bikes > Mountain Bikes         54.0                  0  
1                 0  Bikes > Mountain Bikes         73.0                  0  
2                 0      Bikes > Road Bikes         81.0                  0  
3                 0  Bikes > Mountain Bikes         51.0                  0  
4                 0      Bikes > Road Bikes         52.0                  0  
5                 0      Bikes > Road Bikes         41.0                  0  
6                 0  Bikes > Mountain Bikes         48.0                  0  
7                 0  Bikes > Mountain Bikes         56.0                  0  
8                 0      Bikes > Road Bikes         50.0                  0  
9                 0      Bikes > Road Bikes         53.0                  0  

[10 rows x 31 columns]

---